libmuscle/python/libmuscle/manager/deadlock_detector.py

```diff
@@ -0,0 +1,220 @@
+    """The DeadlockDetector attempts to detect when multiple instances are stuck waiting
+    for each other.
+
+    This class is responsible for handling WAITING_FOR_RECEIVE and
+    WAITING_FOR_RECEIVE_DONE MMP messages, which are submitted by the MMPServer.
+
+    When a deadlock is detected, the cycle of instances that is waiting on each other is
+    logged with FATAL severity. If this deadlock does not get resoled in
+    ``wait_before_shutdown`` seconds, the simulation is shut down.
+    """
+
+    def __init__(
+            self, shutdown_callback: Callable[[], None], wait_before_shutdown: float
+            ) -> None:
+        """Construct a new DeadlockDetector thread.
+
+        Args:
+            shutdown_callback: function to execute when a deadlock is detected. This
+                callback (which is executed in this thread!) is responsible for stopping
+                the simulation when a deadlock is detected.
+            wait_before_shutdown: Number of seconds to wait before executing
+                :param:`shutdown_callback` after a deadlock is detected. If the deadlock
+                is resolved (although this is unlikely), the simulation will not shut
+                down.
+        """
+        super().__init__(name="DeadlockDetector")
+
+        self._shutdown_callback = shutdown_callback
+        self._wait_before_shutdown = wait_before_shutdown
+
+        self._queue: Queue[Optional[_QueueItem]] = Queue()
+        """Queue of incoming messages. Incoming messages can come in any communication
+        thread and will be consumed and processed in this worker thread.
+        """
+        self._waiting_instances: Dict[str, str] = {}
+        """Maps instance IDs to the peer instance IDs they are waiting for."""
+        self._waiting_instance_ports: Dict[str, Tuple[str, Optional[int]]] = {}
+        """Maps instance IDs to the port/slot they are waiting on.."""
+
+        self._detected_deadlocks: List[List[str]] = []
+        """List of deadlocked instance cycles. Set by _handle_potential_deadlock.
+        """
+        self._shutdown_time: Optional[float] = None
+        """Future time when we confirm the potential deadlock and abort the simulation.
+        """
+
+    def run(self) -> None:
+        """Logic that is executed in the thread."""
+        while True:
+            # Set a timeout when a deadlock was detected
+            timeout = None
+            if self._shutdown_time is not None:
+                timeout = max(0, self._shutdown_time - time.monotonic())
+
+            # Grab a new item from the queue, this raises Empty when timeout expires:
+            try:
+                item = self._queue.get(timeout=timeout)
+                if item is None:  # On shutdown, None is pushed to the queue
+                    return  # exit thread
+                self._process_queue_item(item)
+
+            except Empty:
+                # timeout expired and queue is empty, call shutdown callback
```

Comment:
Hold on wait a minute, what if everyone is happily computing, and that's why we're not getting any messages about waits? Or they're communicating but it's all going back and forth so fast that there's never anyone waiting long enough to send a message?

I think we need to check here if there are any deadlocks, and if not, continue the while loop.

Reply:
If everyone is happily communicating then they should have sent a `WAITING_FOR_RECEIVE_DONE` message.
This message would be processed before the timeout expires, and if there are no deadlocks, there is no new timeout set.

It's not the most intuitive logic though, so I'll add a comment and put an assert double-checking for deadlocks here :)

Reply:
Ah, you're right, I didn't read carefully enough how the timeout is set. It makes sense, it'll only specify a timeout if we've already detected a deadlock. So PEBKAC here.

Maybe it would help if we introduced some terminology to more clearly distinguish between the different timeouts. There's a client-side timeout after which the manager is informed, and there's a server-side timeout to compensate for any potential delayed messages. Something with a grace period for the latter?